File: backend/modules/turnament-management/domain/turnament.ts
export type TurnamentState = {
  id: string;
  name: string;
};

export class TurnamentCreated {
  readonly id: string;
  readonly name: string;

  constructor(event: TurnamentCreated) {
    this.id = event.id;
    this.name = event.name;
  }
}

type TurnamentEvents = TurnamentCreated;

export class Turnament {
  private id: string;

  constructor(readonly state: TurnamentState) {
    this.id = state.id;
  }

  static create(state: TurnamentState) {
    const entity = new Turnament(state);
    entity.events.push(new TurnamentCreated(state));
    return entity;
  }

  // Domain events
  private events: TurnamentEvents[] = [];
  getEvents() {
    return this.events;
  }
}

export interface TurnamentRepository {
  load(id: string): Promise<Turnament | undefined>;
  save(turnament: Turnament): Promise<void>;
}

export class InMemoryTurnamentRepository implements TurnamentRepository {
  private turnaments: Turnament[] = [];
  async load(id: string): Promise<Turnament | undefined> {
    const turnament = this.turnaments.find((turnament) => turnament.state.id === id);
    if (!turnament) {
      return undefined;
    }
    return turnament;
  }
  async save(turnament: Turnament): Promise<void> {
    this.turnaments = this.turnaments.filter((t) => t.state.id !== turnament.state.id);
    this.turnaments.push(turnament);
  }
}

File: backend-go/internal/turnament-management/domain/turnament.go
package turnament_management

type TurnamentState struct {
	Id   string
	Name string
}

type Turnament struct {
	State   TurnamentState
	events  []any
}

// events
type TurnamentCreated struct {
	Id   string
	Name string
}

func CreateTurnament(Id string, Name string) *Turnament {
	return &Turnament{
		State:  TurnamentState{Id: Id, Name: Name},
		events: []any{TurnamentCreated{Id: Id, Name: Name}},
	}
}

// GetEvents returns the events of a turnament
func (t *Turnament) GetEvents() []any {
	return t.events
}

type TurnamentRepository interface {
	Load(Id string) (*Turnament, error)
	Save(t *Turnament) error
}

type InMemoryTurnamentRepository struct {
	turnaments []*Turnament
}

func (r *InMemoryTurnamentRepository) Load(Id string) (*Turnament, error) {
	for _, t := range r.turnaments {
		if t.State.Id == Id {
			return t, nil
		}
	}
	return nil, nil
}

func (r *InMemoryTurnamentRepository) Save(t *Turnament) error {
	for i, turnament := range r.turnaments {
		if turnament.State.Id == t.State.Id {
			r.turnaments[i] = t
			return nil
		}
	}
	r.turnaments = append(r.turnaments, t)
	return nil
}
